src/ServiceControl/Upgrade/StaleIndexAfterUpgrade.cs

```diff
@@ -0,0 +1,84 @@
+﻿namespace Particular.ServiceControl.Upgrade
+{
+    using System;
+    using System.IO;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NServiceBus;
+    using NServiceBus.Features;
+    using ServiceBus.Management.Infrastructure.Settings;
+
+    public class StaleIndexAfterUpgrade : Feature
+    {
+        public StaleIndexAfterUpgrade()
+        {
+            EnableByDefault();
+            RegisterStartupTask<CheckerTask>();
+        }
+        
+        protected override void Setup(FeatureConfigurationContext context)
+        {
+            context.Container.ConfigureComponent<StaleIndexChecker>(DependencyLifecycle.SingleInstance);
+            context.Container.ConfigureComponent<StaleIndexInfoStore>(DependencyLifecycle.SingleInstance);
+        }
+        
+        public class CheckerTask : FeatureStartupTask 
+        {
+            private string directory;
+            private Task checkTask;
+            private CancellationTokenSource tokenSource;
+            private StaleIndexChecker staleIndexChecker;
+            private StaleIndexInfoStore staleIndexInfoStore;
+
+            public CheckerTask(StaleIndexChecker staleIndexChecker, StaleIndexInfoStore staleIndexInfoStore, LoggingSettings loggingSettings) : this(staleIndexChecker, staleIndexInfoStore, loggingSettings, null) {}
+            
+            protected CheckerTask(StaleIndexChecker staleIndexChecker, StaleIndexInfoStore staleIndexInfoStore, LoggingSettings loggingSettings, string baseDirectory)
+            {
+                this.staleIndexInfoStore = staleIndexInfoStore;
+                this.staleIndexChecker = staleIndexChecker;
+
+                directory = baseDirectory ?? loggingSettings.LogPath;
+            }
+
+            protected override void OnStop()
+            {
+                tokenSource?.Cancel();
+                checkTask?.GetAwaiter().GetResult();
+            }
+
+            protected override void OnStart()
+            {
+                var strings = Directory.GetFiles(directory, "*.upgrade");
```

Comment:
Is it possible for two instances to share a log folder? Do we think that's a common scenario?

Reply:
The normal log files would be pretty unusable at that stage, I don't think that's a scenario we can support.